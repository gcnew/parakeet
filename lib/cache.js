"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function cache(accessor, leftReplace) {
    return function (p) { return function (st) {
        var pos = st.getPosition();
        var data = st.getData();
        var cache = accessor(data, pos, p);
        if (cache && (cache.kind === 'right' || !leftReplace)) {
            return cache;
        }
        var val = p(st);
        accessor(data, pos, p, val);
        return val;
    }; };
}
exports.cache = cache;
function mkMap() {
    var map = Object.create(null);
    map['del me'] = 1;
    delete map['del me'];
    return map;
}
var EMPTY_MAP = mkMap();
function cacheChain(f) {
    return function (get, set) { return function (data, pos, p, val) {
        var cache = get(data) || (val ? mkMap() : EMPTY_MAP);
        var entry = cache[pos];
        if (val === undefined) {
            if (!entry) {
                return undefined;
            }
            var key_1 = f(p, data);
            if (Array.isArray(entry)) {
                for (var _i = 0, entry_1 = entry; _i < entry_1.length; _i++) {
                    var pair = entry_1[_i];
                    if (pair.key === key_1) {
                        return pair.value;
                    }
                }
                return undefined;
            }
            return entry.key === key_1 ? entry.value
                : undefined;
        }
        var key = f(p, data);
        var kv = { key: key, value: val };
        if (!entry) {
            cache[pos] = kv;
            set(data, cache);
            return;
        }
        var arr = Array.isArray(entry) ? entry : [entry];
        for (var _a = 0, arr_1 = arr; _a < arr_1.length; _a++) {
            var pair = arr_1[_a];
            if (pair.key === key) {
                if (pair.value.kind === 'right') {
                    throw new Error('ASSERT: right replacement encountered');
                }
                pair.value = val;
                return;
            }
        }
        arr.push(kv);
        cache[pos] = arr;
        return;
    }; };
}
exports.byPosAndParser = cacheChain(function (x) { return x; });
function byPosAndKey(get, set) {
    return function (key) { return cacheChain(function (_) { return key; })(get, set); };
}
exports.byPosAndKey = byPosAndKey;
