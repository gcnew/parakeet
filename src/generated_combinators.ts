
import { Parser, ParserStream } from './parser_combinators'

export function genericCombine<M, S extends ParserStream<M>, E, A, B>(
    parsers: Parser<M, S, E, A>[],
    f: (...values: A[]) => B
): Parser<M, S, E, B> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let cur = st;
        const results: any[] = [];

        for (const p of parsers) {
            const res = p(cur);
            if (res.kind === 'left') {
                return res;
            }

            cur = res.value[1];
            results.push(res.value[0]);
        }

        return { kind: 'right', value: [ f(...results), cur ] };
    };
}

export function genericAlt<M, S extends ParserStream<M>, E, T>(parsers: Parser<M, S, E, T>[]): Parser<M, S, E, T> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let err = undefined;
        for (const p of parsers) {
            const res = p(st);
            if (res.kind !== 'left') {
                return res;
            }

            err = res;
        }

        return err!;
    };
}

export function genericChoice<M, S extends ParserStream<M>, E1, E2, T>(
    parsers: [Parser<M, S, E1, unknown>, Parser<M, S, E2, T>][]
): Parser<M, S, E1|E2, T> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x[0] === 'function' && typeof x[1] === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let ret = undefined;
        for (const pair of parsers) {
            const res = pair[0](st);

            if (res.kind === 'right') {
                return pair[1](res.value[1]);
            }

            ret = res;
        }

        return ret!;
    };
}

export function alt<M, S extends ParserStream<M>, E, T1, T2>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, E, T2>
): Parser<M, S, E, T1|T2> {
    return genericAlt<M, S, any, T1|T2>([p1, p2]);
}

export function alt3<M, S extends ParserStream<M>, E, T1, T2, T3>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, E, T3>
): Parser<M, S, E, T1|T2|T3> {
    return genericAlt<M, S, any, T1|T2|T3>([p1, p2, p3]);
}

export function alt4<M, S extends ParserStream<M>, E, T1, T2, T3, T4>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, E, T4>
): Parser<M, S, E, T1|T2|T3|T4> {
    return genericAlt<M, S, any, T1|T2|T3|T4>([p1, p2, p3, p4]);
}

export function alt5<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, E, T5>
): Parser<M, S, E, T1|T2|T3|T4|T5> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5>([p1, p2, p3, p4, p5]);
}

export function alt6<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, E, T6>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6>([p1, p2, p3, p4, p5, p6]);
}

export function alt7<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6, T7>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, unknown, T6>,
    p7: Parser<M, S, E, T7>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6|T7> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6|T7>([p1, p2, p3, p4, p5, p6, p7]);
}

export function alt8<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, unknown, T6>,
    p7: Parser<M, S, unknown, T7>,
    p8: Parser<M, S, E, T8>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6|T7|T8> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6|T7|T8>([p1, p2, p3, p4, p5, p6, p7, p8]);
}

export function combine<M, S extends ParserStream<M>, E1, E2, T1, T2, T3>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    f: (a1: T1, a2: T2) => T3
): Parser<M, S, E1|E2, T3> {
    return genericCombine<M, S, E1|E2, any, T3>([p1, p2], f);
}

export function combine3<M, S extends ParserStream<M>, E1, E2, E3, T1, T2, T3, T4>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    f: (a1: T1, a2: T2, a3: T3) => T4
): Parser<M, S, E1|E2|E3, T4> {
    return genericCombine<M, S, E1|E2|E3, any, T4>([p1, p2, p3], f);
}

export function combine4<M, S extends ParserStream<M>, E1, E2, E3, E4, T1, T2, T3, T4, T5>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4) => T5
): Parser<M, S, E1|E2|E3|E4, T5> {
    return genericCombine<M, S, E1|E2|E3|E4, any, T5>([p1, p2, p3, p4], f);
}

export function combine5<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, T1, T2, T3, T4, T5, T6>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5) => T6
): Parser<M, S, E1|E2|E3|E4|E5, T6> {
    return genericCombine<M, S, E1|E2|E3|E4|E5, any, T6>([p1, p2, p3, p4, p5], f);
}

export function combine6<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, T1, T2, T3, T4, T5, T6, T7>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6) => T7
): Parser<M, S, E1|E2|E3|E4|E5|E6, T7> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6, any, T7>([p1, p2, p3, p4, p5, p6], f);
}

export function combine7<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    p7: Parser<M, S, E7, T7>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6, a7: T7) => T8
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7, T8> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6|E7, any, T8>([p1, p2, p3, p4, p5, p6, p7], f);
}

export function combine8<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, T1, T2, T3, T4, T5, T6, T7, T8, T9>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    p7: Parser<M, S, E7, T7>,
    p8: Parser<M, S, E8, T8>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6, a7: T7, a8: T8) => T9
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8, T9> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6|E7|E8, any, T9>([p1, p2, p3, p4, p5, p6, p7, p8], f);
}

export function choice<M, S extends ParserStream<M>, E1, E2, E3, T1, T2>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, E3, unknown>, Parser<M, S, E2, T2>]
): Parser<M, S, E1|E2|E3, T1|T2> {
    return genericChoice<M, S, any, E1|E2, T1|T2>([p1, p2]);
}

export function choice3<M, S extends ParserStream<M>, E1, E2, E3, E4, T1, T2, T3>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, E4, unknown>, Parser<M, S, E3, T3>]
): Parser<M, S, E1|E2|E3|E4, T1|T2|T3> {
    return genericChoice<M, S, any, E1|E2|E3, T1|T2|T3>([p1, p2, p3]);
}

export function choice4<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, T1, T2, T3, T4>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, E5, unknown>, Parser<M, S, E4, T4>]
): Parser<M, S, E1|E2|E3|E4|E5, T1|T2|T3|T4> {
    return genericChoice<M, S, any, E1|E2|E3|E4, T1|T2|T3|T4>([p1, p2, p3, p4]);
}

export function choice5<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, T1, T2, T3, T4, T5>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, E6, unknown>, Parser<M, S, E5, T5>]
): Parser<M, S, E1|E2|E3|E4|E5|E6, T1|T2|T3|T4|T5> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5, T1|T2|T3|T4|T5>([p1, p2, p3, p4, p5]);
}

export function choice6<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, T1, T2, T3, T4, T5, T6>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, E7, unknown>, Parser<M, S, E6, T6>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7, T1|T2|T3|T4|T5|T6> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6, T1|T2|T3|T4|T5|T6>([p1, p2, p3, p4, p5, p6]);
}

export function choice7<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, T1, T2, T3, T4, T5, T6, T7>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, unknown, unknown>, Parser<M, S, E6, T6>],
    p7: [Parser<M, S, E8, unknown>, Parser<M, S, E7, T7>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8, T1|T2|T3|T4|T5|T6|T7> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6|E7, T1|T2|T3|T4|T5|T6|T7>([p1, p2, p3, p4, p5, p6, p7]);
}

export function choice8<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, E9, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, unknown, unknown>, Parser<M, S, E6, T6>],
    p7: [Parser<M, S, unknown, unknown>, Parser<M, S, E7, T7>],
    p8: [Parser<M, S, E9, unknown>, Parser<M, S, E8, T8>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8|E9, T1|T2|T3|T4|T5|T6|T7|T8> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6|E7|E8, T1|T2|T3|T4|T5|T6|T7|T8>([p1, p2, p3, p4, p5, p6, p7, p8]);
}
