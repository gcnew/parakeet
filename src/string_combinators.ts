
import { Parser, left, right, pair } from './parser_combinators'
import {
    ParserStream, Either,

    combine, combine3, alt, many, trai, satisfy, inspect, pfail, pconst, genericChoice
}  from './parser_combinators'

export {
    TextStream, StringParser, LineOffsetTable, WithPosition, WithLineCol,

    StringParserError, DigitExpected, WhitespaceExpected, UnderscoreExpected,
    AsciiAlphaExpected, AsciiIdCharExpected, StringMismatch, CharNotExpected,
    InvalidPosition,

    oneOf, stringChoice, string,

    position, withPosition, posToLineCol, posToLineCol2,

    parseLineOffsets, getLineCol
}

export type char = string;

interface TextStream extends ParserStream<char> {
    getPosition(this: this): number
}

type StringParser<E, T> = Parser<char, TextStream, E, T>

type LineOffsetTable = [number, string][];

type WithPosition<T> = { position: number, value: T }
type WithLineCol<T>  = { line: number, col: number, value: T }

type DigitExpected       = { kind: 'pc_error', code: 'digit_expected'         }
type WhitespaceExpected  = { kind: 'pc_error', code: 'whitespace_expected'    }
type UnderscoreExpected  = { kind: 'pc_error', code: 'underscore_expected'    }
type AsciiAlphaExpected  = { kind: 'pc_error', code: 'ascii_alpha_expected'   }
type AsciiIdCharExpected = { kind: 'pc_error', code: 'ascii_id_char_expected' }

type StringParserError = DigitExpected
                       | WhitespaceExpected
                       | UnderscoreExpected
                       | AsciiAlphaExpected
                       | AsciiIdCharExpected
                       | StringMismatch
                       | CharNotExpected

type StringMismatch    = { kind: 'pc_error', code: 'string_mismatch',   expected: string }
type CharNotExpected   = { kind: 'pc_error', code: 'char_not_expected', expected: string }
type InvalidPosition   = { kind: 'pc_error', code: 'invalid_position'                    }


export const ws          = charParser(isWhiteSpace,  mkSimpleError('whitespace_expected'));
export const digit       = charParser(isDigit,       mkSimpleError('digit_expected'));
export const under       = charParser(isUnder,       mkSimpleError('underscore_expected'));
export const asciiAlpha  = charParser(isAsciiAlpha,  mkSimpleError('ascii_alpha_expected'));
export const asciiIdChar = charParser(isAsciiIdChar, mkSimpleError('ascii_id_char_expected'));

export const char = string;

export const asciiId = combine(
    alt(under, asciiAlpha),
    many(asciiIdChar),
    (x, xs) => x + xs.join('')
);

export const integer = combine(
    digit,
    many(digit),
    (x, xs) => x + xs.join('')
);

export const float = combine3(
    integer,
    char('.'),
    integer,
    (i1, _, i2) => i1 + '.' + i2
);

export const number = combine(
    integer,
    trai(char('.'), integer, (_, i) => '.' + i),
    (w, f) => w + (f || '')
);

function string<S extends string>(s: S): StringParser<StringMismatch, S> {
    if (!s.length) {
        throw new Error('Empty string');
    }

    const errorData: StringMismatch = { kind: 'pc_error', code: 'string_mismatch', expected: s };
    const error = left(errorData);

    return (st) => {
        let cur = st;

        for (let i = 0; i < s.length; ++i) {
            const next = cur.next();

            if (!next) {
                return error;
            }

            if (next[0] !== s[i]) {
                return error;
            }

            cur = next[1];
        }

        return right(pair(s, cur));
    };
}

function oneOf(s: string): StringParser<CharNotExpected, char> {
    if (!s.length) {
        throw new Error('Empty string');
    }

    const charMap: { [key: string]: boolean } = {};
    for (let c of s) {
        charMap[c] = true;
    }

    const errorData: CharNotExpected = { kind: 'pc_error', code: 'char_not_expected', expected: s };
    const error = left(errorData);

    return (st) => {
        const next = st.next();

        if (!next || !charMap[next[0]]) {
            return error;
        }

        return right(next);
    };
}

// TODO: one day
// function choice<T, M extends { [key: string]: StringParser<any, T> }>(map: M): StringParser<ErrorOf<M[keyof M]> | StringMismatch, T> {

function stringChoice<E, T>(map: { [key: string]: StringParser<E, T> }): StringParser<E | StringMismatch, T> {
    const keys = Object.keys(map).sort().reverse();
    return genericChoice(keys.map(k => pair(string(k), map[k])));
}

function position<S extends TextStream>(st: S): Either<never, [number, S]> {
    return right(pair(st.getPosition(), st));
}

function withPosition<E, T>(p: StringParser<E, T>): StringParser<WithPosition<E>, WithPosition<T>> {
    return (st) => {
        const res = p(st);
        const position = st.getPosition();

        if (res.kind === 'left') {
            return left({ position, value: res.value });
        }

        return right(pair({ position, value: res.value[0] }, res.value[1]));
    };
}

const invalid_position: InvalidPosition = { kind: 'pc_error', code: 'invalid_position' };
const left_invalid_position  = left(invalid_position);
const pfail_invalid_position = pfail(invalid_position);

function posToLineCol(table: LineOffsetTable) {
    return <S extends TextStream>(st: S): Either<InvalidPosition, [[number, number], S]> => {
        const pos = st.getPosition();
        const lineCol = getLineCol(pos, table);

        if (!lineCol) {
            return left_invalid_position;
        }

        return right(pair(lineCol, st));
    };
}

function posToLineCol2<E>(pt: StringParser<E, LineOffsetTable>): StringParser<E | InvalidPosition, [number, number]> {
    return inspect(
        combine(pt, position, (t, p) => getLineCol(p, t)),

        lineCol => {
            if (!lineCol) {
                return pfail_invalid_position;
            }

            return pconst(lineCol);
        }
    );
}

function getLineCol(offset: number, lineOffsetTable: LineOffsetTable): [number, number]|null {
    let idx = binarySearch(lineOffsetTable, x => x[0] < offset ? -1 :
                                                 x[0] > offset ?  1 : 0);

    if (idx === false) {
        return null;
    }

    if (idx < 0) {
        idx = -idx - 1;
    }

    return [idx, offset - lineOffsetTable[idx][0]];
}

function parseLineOffsets(source: string): LineOffsetTable {
    const lines = source.split('\n');

    let acc = [];
    let offset = 0;
    for (const l of lines) {
        acc.push(pair(offset, l));
        offset += l.length + 1;
    }

    return acc;
}

function binarySearch<T>(arr: T[], compare: (x: T) => -1|0|1): false|number {
    let low = 0;
    let high = arr.length - 1;

    if (!arr.length) {
        return false;
    }

    while (low <= high) {
        const m = low + ((high - low) >> 1);
        const cmp = compare(arr[m]);

        if (cmp < 0) {
            low = m + 1;
            continue;
        }

        if (cmp > 0) {
            high = m - 1;
            continue;
        }

        return m;
    }

    return -low;
}

function isWhiteSpace(s: char) {
    switch (s) {
        case ' ': case '\t': case '\n': case '\r':
            return true;
        default:
            return false;
    }
}

function isDigit(s: char) {
    return s >= '0' && s <= '9';
}

function isUnder(s: char) {
    return s === '_';
}

function isAsciiAlpha(s: char) {
    return (s >= 'a' && s <= 'z') || (s >= 'A' && s <= 'Z');
}

function isAsciiIdChar(s: char) {
    return isAsciiAlpha(s) || isUnder(s) || isDigit(s);
}

function mkSimpleError<T extends string>(code: T) {
    return { kind: 'pc_error' as 'pc_error', code };
}

function charParser<E extends StringParserError>(pred: (x: char) => boolean, error: E): StringParser<E, char> {
    return satisfy<char, TextStream, E>(pred, error);
}
