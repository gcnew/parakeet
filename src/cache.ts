
import { StringParser } from './string_combinators'

export {
    Accessor,

    cache, byPosAndKey
}


type Accessor = (
        data: any,
        position: number,
        parser: StringParser<any, any>,
        value?: any
    ) => any;


function cache(accessor: Accessor, leftReplace: boolean): <E, T>(p: StringParser<E, T>) => StringParser<E, T> {
    return <E, T>(p: StringParser<E, T>): StringParser<E, T> => st => {
        const pos = st.getPosition();
        const data = st.getData();
        const cache = accessor(data, pos, p);

        if (cache && (cache.kind === 'right' || !leftReplace)) {
            return cache;
        }

        const val = p(st);

        // WARNING: mutation - set the value in the state
        accessor(data, pos, p, val);

        return val;
    };
}

function mkMap() {
    const map: { [key: string]: any } = Object.create(null);

    map['del me'] = 1;
    delete map['del me'];

    return map;
}

const EMPTY_MAP = mkMap();

function cacheChain(f: (parser: StringParser<any, any>, data: any) => any):
    (
        get: (data: any) => any,
        set: (data: any, value: any) => void
    ) => Accessor
{
    return (get, set) => (data, pos, p, val) => {
        const cache = get(data) || (val ? mkMap() : EMPTY_MAP);
        const entry = cache[pos];

        // getter
        if (val === undefined) {
            if (!entry) {
                return undefined;
            }

            const key = f(p, data);
            if (Array.isArray(entry)) {
                for (const pair of entry) {
                    if (pair.key === key) {
                        return pair.value;
                    }
                }

                return undefined;
            }

            return entry.key === key ? entry.value
                                     : undefined;
        }

        // setter
        const key = f(p, data);
        const kv = { key, value: val };
        if (!entry) {
            cache[pos] = kv;
            set(data, cache);
            return;
        }

        const arr = Array.isArray(entry) ? entry : [entry];
        for (const pair of arr) {
            if (pair.key === key) {
                if (pair.value.kind === 'right') {
                    throw new Error('ASSERT: right replacement encountered');
                }

                pair.value = val;
                return;
            }
        }

        arr.push(kv);
        cache[pos] = arr;
        return;
    };
}

export const byPosAndParser = cacheChain(x => x);

function byPosAndKey(
    get: (data: any) => any,
    set: (data: any, value: any) => any
): (key: string) => Accessor {
    return key => cacheChain(_ => key)(get, set);
}
