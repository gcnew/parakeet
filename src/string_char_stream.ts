
import { pair } from './parser_combinators'
import { TextStream } from './string_combinators'

export { StringCharStream }

type LineOffsetTable = [number, string][];

class StringCharStream implements TextStream {
    protected constructor(
        private source: string,
        private offset: number,
        private lineOffsetTable: [number, string][]
    ) {
    }

    protected newInstance(source: string, offset: number, lineOffsetTable: [number, string][]): this {
        return new StringCharStream(source, offset, lineOffsetTable) as this;
    }

    public static newInstance(source: string) {
        return new StringCharStream(source, 0, parseLineOffsets(source));
    }

    next() {
        if (this.offset >= this.source.length) {
            return null;
        }

        return pair(
            this.source[this.offset],
            this.newInstance(this.source, this.offset + 1, this.lineOffsetTable)
        );
    }

    getPosition() {
        return this.offset;
    }

    getLineCol(offset: number): [number, number] {
        const res = getLineCol(offset, this.lineOffsetTable);

        if (!res) {
            throw new Error(`Invalid offset: ${offset}`);
        }

        return res;
    }
}

function getLineCol(offset: number, lineOffsetTable: LineOffsetTable): [number, number]|null {
    let idx = binarySearch(lineOffsetTable, x => x[0] < offset ? -1 :
                                                 x[0] > offset ?  1 : 0);

    if (idx === false) {
        return null;
    }

    if (idx < 0) {
        idx = -idx - 1;
    }

    return [idx, offset - lineOffsetTable[idx][0]];
}

function parseLineOffsets(source: string): LineOffsetTable {
    const lines = source.split('\n');

    let acc = [];
    let offset = 0;
    for (const l of lines) {
        acc.push(pair(offset, l));
        offset += l.length + 1;
    }

    return acc;
}

function binarySearch<T>(arr: T[], compare: (x: T) => -1|0|1): false|number {
    let low = 0;
    let high = arr.length - 1;

    if (!arr.length) {
        return false;
    }

    while (low <= high) {
        const m = low + ((high - low) >> 1);
        const cmp = compare(arr[m]);

        if (cmp < 0) {
            low = m + 1;
            continue;
        }

        if (cmp > 0) {
            high = m - 1;
            continue;
        }

        return m;
    }

    return -low;
}
