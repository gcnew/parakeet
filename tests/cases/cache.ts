
import { test } from '../test_util'

import * as C from '../../src/cache'

import {
    StringParser,

    char, string
} from '../../src/string_combinators'

import {
    map, alt3, combine, combine3, getData, pair
} from '../../src/parser_combinators'


const cache1 = C.cache(
    C.byPosAndParser(
        data => data.cache,
        (data, cache) => data.cache = cache
    ),
    false
);

const cache2 = (key: string, replace: boolean) => C.cache(
    C.byPosAndKey(
        data => data.cache,
        (data, cache) => data.cache = cache
    )(key),
    replace
);


function mkParser(cache: typeof cache1) {
    const x = cache(combine(char('x'), getData, (x, d) => (d.counter++, x)));

    const xn = alt3(
        string('pasta'),
        combine3(x, x, cache(char('y')), (a, b, y) => a + b + y),
        combine3(x, x, x,                (a, b, c) => a + b + c)
    );

    return combine(
        map(getData, d => d.counter = 0) as StringParser<never, 0>,
        combine(xn, getData, (r, d) => pair(r, d.counter)),

        (_, x) => x
    );
}

test(mkParser(cache1), 'xxx');
test(mkParser(cache2('eks', true)), 'xxx');

// Failure expected as the fail of `char('y')` is persisted and returned
test(mkParser(cache2('eks', false)), 'xxx');
