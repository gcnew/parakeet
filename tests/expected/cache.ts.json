{
    "kind": "right",
    "value": [
        [
            "xxx",
            3
        ],
        {
            "source": "xxx",
            "offset": 3,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "xxx"
                    ]
                ],
                "counter": 3,
                "cache": {
                    "0": {
                        "value": {
                            "kind": "right",
                            "value": [
                                "x",
                                {
                                    "source": "xxx",
                                    "offset": 1,
                                    "data": "Cycle(root.value[1].data.cache[\"0\"].value.value[1].data)"
                                }
                            ]
                        }
                    },
                    "1": {
                        "value": {
                            "kind": "right",
                            "value": [
                                "x",
                                {
                                    "source": "xxx",
                                    "offset": 2,
                                    "data": "Cycle(root.value[1].data.cache[1].value.value[1].data)"
                                }
                            ]
                        }
                    },
                    "2": [
                        {
                            "value": {
                                "kind": "left",
                                "value": {
                                    "kind": "pc_error",
                                    "code": "string_mismatch",
                                    "expected": "y"
                                }
                            }
                        },
                        {
                            "value": {
                                "kind": "right",
                                "value": [
                                    "x",
                                    "Cycle(root.value[1].data.cache[2][1].value.value[1])"
                                ]
                            }
                        }
                    ]
                }
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "xxx",
            3
        ],
        {
            "source": "xxx",
            "offset": 3,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "xxx"
                    ]
                ],
                "counter": 3,
                "cache": {
                    "0": {
                        "key": "eks",
                        "value": {
                            "kind": "right",
                            "value": [
                                "x",
                                {
                                    "source": "xxx",
                                    "offset": 1,
                                    "data": "Cycle(root.value[1].data.cache[\"0\"].value.value[1].data)"
                                }
                            ]
                        }
                    },
                    "1": {
                        "key": "eks",
                        "value": {
                            "kind": "right",
                            "value": [
                                "x",
                                {
                                    "source": "xxx",
                                    "offset": 2,
                                    "data": "Cycle(root.value[1].data.cache[1].value.value[1].data)"
                                }
                            ]
                        }
                    },
                    "2": {
                        "key": "eks",
                        "value": {
                            "kind": "right",
                            "value": [
                                "x",
                                "Cycle(root.value[1].data.cache[2].value.value[1])"
                            ]
                        }
                    }
                }
            }
        }
    ]
}

{
    "kind": "left",
    "value": {
        "kind": "pc_error",
        "code": "string_mismatch",
        "expected": "y"
    }
}
